package.json

```diff
@@ -262,6 +262,7 @@
     "prettyjson": "^1.1.3",
     "progress-stream": "^2.0.0",
     "reconfix": "^1.0.0-v0-1-0-fork-46760acff4d165f5238bfac5e464256ef1944476",
+    "recursive-fs": "^2.1.0",
```

Comment:
Something I should have remembered before, but better late than never... The current implementation using `klaw` is something we arrived at in PR #2093 from @joshbwlng (a year ago!) that was aimed at improving performance for Jellyfish with livepush -- not speed, but reducing high CLI memory usage of a previous ***recursive*** directory traversal in a custom `listFiles()` function:

> https://github.com/balena-io/balena-cli/pull/2093#issuecomment-727771295
> Been testing this branch today and it looks like there might be another problem that causes `filterFilesWithDockerignore()` to fail when too many files exist under the project root. Jellyfish currently has multiple `node_modules` directories, five when all `npm install`s have been ran (one for the project root along with one for each app). After adding a few debug logs, I was able to track down the problem to the `listFiles()` function in `lib/utils/ignore.ts`. This runs before dockerignore patterns are taken into account, also causing non-livepush `push`es to fail during `newTarDirectory()`. 

The previous recursive traversal in a `listFiles()` function (replaced with `klaw`) can be seen in commit: [push: Reduce memory usage when filtering files with dockerignore](https://github.com/balena-io/balena-cli/pull/2093/commits/4768f763856aa9f761988477f97ec872d226b004).

That previous `listFiles()` function probably had higher memory usage than the use of `recursive-fs` proposed in this PR, I think mainly because `listFiles()` executed `fs.stat` on _all files before filtering,_ which this PR cleverly avoids. Nonetheless, `recursive-fs` also performs _recursive_ directory traversal (the hint is in the package name...), as we see in the following `walk` function that calls itself:
https://github.com/simov/recursive-fs/blob/v2.1.0/lib/read.js#L30

I leave it to you to decide, but the thought that occurred to me (out of the blue while having a shower!), compounded by the fact that `recursive-fs` has the forward slash vs `path.join()` Windows issue, is to replace `recursive-fs` with a custom `listFiles()` function of a year ago, with a few key changes:

* The new `listFiles()` function would **not** call `fs.stat` on all files before filtering, which I understand is the key speed improvement of this PR.
* The new `listFiles()` function would have an iterative, **non-recursive** implementation, hoping that this saves some memory by avoiding the automated function call stack build-up. Ref: [Recursion to Iteration](https://blog.moertel.com/posts/2013-05-11-recursive-to-iterative.html)
* Like the old `listFiles()` function and like `recursive-fs`, the new `listFiles()` function would be based on `fs.readdir()`, and would take advantage of `entry.isDirectory()` and other [fs.Dirent](https://nodejs.org/docs/latest-v12.x/api/fs.html#fs_class_fs_dirent) methods without needing a call to `fs.stat`.

Reply:
Thanks for all the extra information @pdcastro 
When you were looking to reduce memory, how were you testing memory consumption? I've been developing the solution using Jellyfish as a test bed, which generates a list of ~277k files, but I have quite a powerful workstation so don't really see any issues with memory.
I'd quite like to avoid re-inventing a file scanning script, so if we can find an existing solution that works that would be the best solution. 
As an additional thought, tarballing a directory for consumption by Docker seems to be a generally useful tool that could exist outside of the CLI - what are your thoughts on this?

Reply:
I also think that the big culprit of memory consumption was the call to `stat` rather than the recursive programming style.

Reply:
> how were you testing memory consumption?

I don't remember having actually measured it (embarrassingly). 😊  I think we had isolated CLI versions where `push` worked and where it didn't, we found that `listFiles()` is something that had changed between those, and found that using `klaw` avoided the failure reported by Josh.

Re not re-inventing a file scanning script, yes I agree. 👍  At the same time, I don't feel strongly between the options of adding a small function to enumerate files around `fs.readdir()`, versus adding a new dependency to `package.json`. Adding dependencies also adds some maintenance cost over time, for example we regularly need to update dependencies (bumping them in `package.json`) for reasons such as updating the supported versions of Node.js or addressing `npm audit` concerns (because of changes made by 3rd parties that may or may not benefit us), and when doing so we need to check the dependencies' changelog for what's changed, especially new major versions.

> tarballing a directory for consumption by Docker seems to be a generally useful tool that could exist outside of the CLI - what are your thoughts on this?

Related packages that come to mind:

* Our own `tar-utils` package:
    * https://www.npmjs.com/package/tar-utils
    * https://github.com/balena-io-modules/node-tar-utils/
* [tar-stream](https://www.npmjs.com/package/tar-stream)
* [@balena/dockerignore](https://www.npmjs.com/package/@balena/dockerignore)
* Our own `docker-toolbelt`
    * https://www.npmjs.com/package/docker-toolbelt
    * https://github.com/balena-io-modules/docker-toolbelt
* [dockerode](https://www.npmjs.com/package/dockerode)

I think you were thinking about a new package that combined tar-stream with dockerignore. Our `docker-toolbelt` package could be a candidate for hosting the functionality. (`docker-toolbelt` could do with a full rewriting, not just converting from Coffeescript to Typescript but also a review of the functionality it provides and the public interface.)

A complexity that the module would have to address is where to expect the `.dockerignore` file(s) to be, i.e. the difference between providing and not providing the balena CLI's `--multi-dockerignore (-m)` flag, and compatibility with the "standard" `docker-compose build` and `docker build` behaviour.